### PR TITLE
feat(boss-chain): thread chain spec to child gremlins as north-star context

### LINKED ISSUE #194: Thread chain spec through to child gremlin implement stage
# Thread the chain spec to child gremlins as read-only context

*Backward compatibility is not a concern — break callers, on-disk
formats, or prompt shapes freely if it simplifies the result. Look
for opportunities to simplify and reuse existing code; prefer
existing helpers in `gremlins/launcher.py`, `gremlins/state.py`,
the implement-stage prompt rendering, etc., over reimplementation.
Keep prompts and code as simple as possible — added complexity now
makes future changes worse.*

## Background

When a boss chain hands work to a child gremlin, the child only
sees the sliced child plan that `handoff.py` produced. The
overarching spec — the original `--plan <spec>` the boss is
chasing — is **not** passed through. The child has no way to know
what shape the chain as a whole is supposed to land, just its
own slice.

Result: children produce partial code. They see "Task: change X to
Y" and miss surrounding callers, downstream consumers, or
follow-on edits that a glance at the spec would have flagged. The
slice was meant to prevent overlap *between* children, but in
practice it also blinds each child to the context it needs to
make coherent local decisions.

The handoff agent already gets exactly this kind of context: it
takes `--spec <path>` as a read-only "north star" the rolling plan
diverges from (`gremlins/handoff.py:113-125`). The fix is to do
the same thing one layer down — give the child gremlin the same
read-only north star, with prompt language scoping it to context
only (no expanding the task list).

## Goal

After this slice:

- Boss `launch_child` passes the chain's spec snapshot through to
  the child localgremlin / ghgremlin via the launcher.
- The launcher accepts a `spec_path` parameter and threads it into
  the child's `pipeline_args` as `--spec <abs-path>`, so it
  survives a resume.
- The local + gh orchestrators accept `--spec <path>` and feed
  the spec text into the implement stage.
- The implement stage's prompt renders an explicit "North star —
  read-only context, do not expand scope" block above the plan
  text, mirroring the shape used by `handoff.py` already.
- Child gremlins launched directly (no boss, no `--spec`) continue
  to behave exactly as today — the new field is optional throughout.
- A 2-step boss chain demonstrably forwards the spec to each
  child, and the rendered implement prompt contains the north-star
  block.

## Design — same shape the handoff already uses

The handoff agent's prompt (`handoff.py:113-125`) wraps the spec
in a labeled block titled "Overarching goal (north star)" with
explicit instructions: read-only context, do not echo into the
output, used only to scope the next step coherently. We lift that
exact pattern into the child implement prompt.

Key constraints to preserve:

- **Spec is optional everywhere.** Direct invocations of
  `localgremlin` / `ghgremlin` (no boss) don't have a spec; the
  prompt simply omits the north-star block in that case.
- **Path durability.** The boss already snapshots its spec to
  `<boss-state-dir>/spec.md` at chain start
  (`boss.py:_resolve_plan_source`). That snapshot is a stable
  absolute path; the child reads it directly. No new copy is
  needed — the boss state dir lives at least as long as the
  child does.
- **Resume survives.** Threading `--spec <abs-path>` through the
  launcher's `pipeline_args` puts it in the child's
  `state.json` automatically, so `launcher.resume()` rehydrates it
  the same way it does `--plan`.
- **Best-effort read.** Mirror `handoff.py:346-361`: if the spec
  path is missing, unreadable, or empty at read time, warn and
  continue without the north-star block. A vanished spec must not
  crash a running child.

The truncation budget (`50000` chars in `handoff.py:115-116`) is
a sensible default to copy verbatim — the handoff agent uses it,
and the same prompt budget concerns apply to the implement
agent.

### `gremlins/launcher.py`: accept and forward `--spec`

Add `spec_path: Optional[str] = None` to `launch()`. When set:

- Validate it points to an existing non-empty file (raise
  `ValueError` on miss, like the existing `--plan` validation at
  `launcher.py:207-211`).
- Normalize to absolute path (mirror `launcher.py:214-215`).
- Inject `["--spec", spec_path]` at the front of
  `stored_pipeline_args` next to where `--plan` is injected
  (`launcher.py:237-239`).

That's the entire launcher change. Resume gets it for free
because `pipeline_args` is rehydrated by `launcher.resume()`
verbatim.

### `gremlins/orchestrators/boss.py`: pass spec_path through

`launch_child` (`boss.py:347-364`) currently passes
`project_root` and `base_ref` to `launcher.launch()`. Add one
more:

```python
spec_path = boss_state.get("spec_path") or None
child_id = _launch(
    launch_kind,
    plan=child_plan,
    parent_id=gr_id,
    project_root=project_root,
    base_ref=base_ref,
    spec_path=spec_path,
)
```

`boss_state["spec_path"]` is already the snapshot
(`init_boss_state` at `boss.py:200-225`), so no new state field
is needed.

### `gremlins/orchestrators/local.py`: argparse + thread to implement

`_parse_local_args` (`local.py:78-114`):

- Add `parser.add_argument("--spec", dest="spec_path", default=None)`.

`local_main` (`local.py:117-275`):

- After the `--plan` staging block at `local.py:142-149`, add
  symmetric handling for `--spec`. Either snapshot the spec into
  `session_dir / "spec.md"` (mirroring the plan-staging shape so
  rescue is fully self-contained) **or** read the source file
  just-in-time at the implement stage. Pick the simpler form —
  the plan-snapshot pattern is the existing precedent so prefer
  matching it unless it adds real complexity here.
- Read `spec_text` from the staged file when entering
  `stage_implement`. Empty string when no `--spec` was supplied.
- Pass `spec_text=spec_text` to `run_implement_stage(...)` at
  `local.py:223-231`.

### `gremlins/orchestrators/gh.py`: argparse + thread to implement

`_parse_gh_args` (`gh.py:82-113`):

- Add `parser.add_argument("--spec", dest="spec_path", default=None)`.

In the gh orchestrator's main flow, follow the same pattern as
local: read the spec text (either snapshotted or just-in-time) and
pass it to the implement stage call. The gh implement stage
already accepts `kwargs` for the prompt rendering — extend the
call site to pass `spec_text`.

### `gremlins/stages/implement.py`: render the north-star block

`run_implement_stage` (`implement.py:79-156`):

- Add `spec_text: str = ""` to the keyword args of
  `run_implement_stage` *and* `_run_implement_gh`.
- Build a `spec_block` string when `spec_text` is non-empty,
  using the same shape as `handoff.py:113-125`:

  ```python
  def _render_spec_block(spec_text: str) -> str:
      if not spec_text or not spec_text.strip():
          return ""
      body = spec_text[:50000]
      trunc = ""
      if len(spec_text) > 50000:
          trunc = f"\n(spec truncated to 50000 chars; {len(spec_text)} chars total)"
      return (
          "## Overarching goal (north star)\n\n"
          "This is the original chain spec. It is read-only context for\n"
          "understanding what the chain as a whole is working toward. Use\n"
          "it to make coherent local decisions while implementing the plan\n"
          "below — not as a task list. Do not expand scope beyond the\n"
          "Tasks in the plan.\n\n"
          f"~~~~\n{body}\n~~~~{trunc}\n\n"
      )
  ```

  When `spec_text` is empty / whitespace, the block is the empty
  string — no placeholder, no header, no leftover whitespace.

### `gremlins/prompts/implement_local.md` and `implement_gh.md`

Add a `{spec_block}` placeholder above the existing plan / issue
body. Concretely, in `implement_local.md`:

```
When writing code, follow these principles:

{core_principles}

{spec_block}{plan_text}

Implement every task in the plan above by editing code in this repo. ...
```

And matching change in `implement_gh.md`. The block is a
self-contained section ending in two trailing newlines, so when
empty the prompt collapses cleanly to its current shape (no extra
blank lines, no orphan headers).

### Existing tests to update

Any test that constructs `run_implement_stage(...)` directly — add
the new kwarg. Search for it. Fixtures keyed off the implement
prompt body (full-string matching) will need to either widen to
substring matching or include the empty-spec form.

### New: spec rendering unit test

Add to `tests/test_stages_implement.py` (or wherever the implement
stage tests live):

```python
def test_implement_renders_spec_block_when_present(...):
    """When spec_text is non-empty, the implement prompt contains
    the north-star block above the plan text."""
    fake = FakeClaudeClient(...)
    run_implement_stage(
        client=fake,
        spec_text="overall spec body",
        plan_text="task 1: ...",
        ...
    )
    prompt = fake.calls[0].prompt
    assert "Overarching goal (north star)" in prompt
    assert "overall spec body" in prompt
    assert prompt.index("overall spec body") < prompt.index("task 1")

def test_implement_omits_spec_block_when_absent(...):
    """When spec_text is empty, the prompt has no north-star header
    and no extra blank lines."""
    fake = FakeClaudeClient(...)
    run_implement_stage(client=fake, spec_text="", plan_text="...", ...)
    prompt = fake.calls[0].prompt
    assert "Overarching goal" not in prompt
```

### New: launcher spec_path forwarding test

Add to `tests/test_launcher.py`:

```python
def test_launch_threads_spec_path_into_pipeline_args(...):
    """launcher.launch(spec_path=<abs>) puts --spec <abs> into
    state.json's pipeline_args, so resume rehydrates it."""
    spec = tmp_path / "spec.md"
    spec.write_text("the spec")
    gr_id = launch("localgremlin", plan=str(plan), spec_path=str(spec), ...)
    state = json.loads((state_dir / "state.json").read_text())
    assert "--spec" in state["pipeline_args"]
    idx = state["pipeline_args"].index("--spec")
    assert state["pipeline_args"][idx + 1] == str(spec.resolve())

def test_launch_rejects_missing_spec_path(...):
    """spec_path that doesn't exist raises ValueError before any
    state-dir setup."""
    with pytest.raises(ValueError, match="--spec"):
        launch("localgremlin", plan=..., spec_path="/nonexistent")
```

### New: boss-child spec passthrough test

Add to `tests/test_orchestrator_boss.py`:

```python
def test_launch_child_forwards_spec_path(...):
    """When boss_state has spec_path set, boss.launch_child passes
    it as spec_path= to launcher.launch."""
    # Stub launcher.launch; capture kwargs.
    # Set boss_state["spec_path"] = "/path/to/spec.md".
    # Invoke launch_child.
    # Assert captured kwargs["spec_path"] == "/path/to/spec.md".
```

## Verification

- `cd gremlins && PYTHONPATH=. python -m pytest` passes.
- Direct `/localgremlin --plan <plan>` (no boss) still runs; the
  rendered implement prompt has no "Overarching goal" header.
- Boss chain run end-to-end: each child's
  `<child-state-dir>/stream-implement.jsonl` shows a prompt
  starting with the north-star block, scoped to the
  passthrough-only language.
- Child resume after a stop: `launcher.resume()` rehydrates
  `--spec` from `pipeline_args`; implement re-renders with the
  spec block intact.
- `grep -rn "spec_block\|Overarching goal" gremlins/prompts/
  gremlins/stages/implement.py` shows exactly one source-of-truth
  for the block content.

## Non-goals

- Threading the spec into `review-code`, `address-code`,
  `ghreview`, or `ghaddress`. Those stages might benefit too, but
  this slice is scoped to the partial-implementation symptom the
  user reported. Once the seam exists, extending it to other
  stages is mechanical and can be its own slice.
- Adding a `--rolling-plan` flag (the rolling plan is a separate
  document that lists *all* remaining tasks across future
  children). Sharing the rolling plan would compound the overlap
  risk that prompted the slicing in the first place. Re-evaluate
  only if the spec alone proves insufficient.
- Snapshotting the spec into a *new* location. The boss already
  snapshots into `<boss-state-dir>/spec.md`; the child reads
  that absolute path. Whether to also snapshot into
  `<child-state-dir>/spec.md` for full self-containment is a
  judgment call inside `local.py` / `gh.py` (mirror plan-staging
  if simple, skip otherwise) — but no new state field or new
  on-disk schema is added either way.
- Renaming or restructuring the existing implement prompt
  templates beyond adding `{spec_block}`. The placeholder is a
  surgical insertion, not a rewrite.
- Touching `handoff.py`. It already does the right thing; this
  slice copies its shape one layer down.

### PR DESCRIPTION
**Pull request by xbrianh**
## Summary

- Boss `launch_child` now forwards `spec_path` from `boss_state` to `launcher.launch`, which injects `--spec` into `stored_pipeline_args` so resume rehydrates it automatically
- Local and gh orchestrators accept `--spec`, snapshot it into `session_dir/spec.md` on first launch (reusing the snapshot on resume), and pass `spec_text` into `run_implement_stage`
- The implement stage renders an "Overarching goal (north star)" block above the plan when `spec_text` is non-empty, mirroring the shape `handoff.py` already uses; direct invocations without `--spec` continue to behave unchanged

## Test plan

- [ ] `test_launch_threads_spec_path_into_pipeline_args` — verifies `--spec` lands in `state.json pipeline_args`
- [ ] `test_launch_rejects_missing_spec_path` / `test_launch_rejects_empty_spec_path` — error-path validation
- [ ] `test_launch_child_forwards_spec_path` / `test_launch_child_no_spec_path_when_absent` — boss passthrough
- [ ] `test_render_spec_block_*` — unit tests for truncation, empty, whitespace cases
- [ ] `test_implement_renders_spec_block_when_present` / `test_implement_omits_spec_block_when_absent` — prompt rendering

Closes #194

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: gremlins/stages/implement.py [hunk 1]
```diff
@@ -76,6 +76,24 @@ def changes_outside_git(sentinel: pathlib.Path, session_dir: pathlib.Path) -> bo
     return False
 
 
+def _render_spec_block(spec_text: str) -> str:
+    if not spec_text or not spec_text.strip():
+        return ""
+    body = spec_text[:50000]
+    trunc = ""
+    if len(spec_text) > 50000:
```
Comment (xbrianh) [inline gremlins/stages/implement.py:83]: `body = spec_text[:50000]` slices at a hard character boundary, which may cut in the middle of a word or sentence. Since the purpose of this block is orientation context (not verbatim reproduction), slicing at the last newline before the limit would produce a cleaner truncation point:

```python
body = spec_text[:50000]
if len(spec_text) > 50000:
    # prefer a clean line boundary
    cut = spec_text.rfind('\n', 0, 50000)
    body = spec_text[:cut] if cut > 0 else spec_text[:50000]
```

Not a correctness issue — the model handles mid-sentence truncation fine — but it's easy to improve.

File: gremlins/orchestrators/gh.py [hunk 2]
```diff
@@ -284,9 +285,20 @@ def gh_main(argv: List[str], *, client: Optional[ClaudeClient] = None) -> int:
...
 
+    # --spec staging: snapshot into session_dir/spec.md on first launch.
+    # On resume, reuse the existing snapshot (rescue-determinism).
+    if args.spec_path and not spec_file.exists():
+        spec_src = pathlib.Path(args.spec_path)
+        if not spec_src.is_file():
+            die(f"--spec: file not found: {args.spec_path}")
+        if spec_src.stat().st_size == 0:
+            die(f"--spec: file is empty: {args.spec_path}")
+        shutil.copyfile(spec_src, spec_file)
```
Comment (xbrianh) [inline gremlins/orchestrators/gh.py:303]: The file-exists / size-zero checks here are redundant: `launcher.py` already validates and normalizes `spec_path` to absolute before spawning the subprocess. The guard is only reachable on a direct (non-launcher) invocation of the orchestrator. That's fine, but it's worth a comment explaining why the check exists — or dropping it and relying on `shutil.copyfile` to raise a clear `FileNotFoundError` if the file vanishes between launch and start.

**Review by xbrianh — COMMENTED** [review]
Clean implementation with solid test coverage. The end-to-end threading — launcher → boss_state → launch_child → orchestrator → implement stage — is correct, and the rescue/resume invariant (snapshot on first launch, reuse on resume) is handled consistently in both orchestrators. One cosmetic point and one minor style note below; nothing blocking.

File: gremlins/orchestrators/gh.py [hunk 3]
```diff
@@ -88,6 +88,7 @@ def _parse_gh_args(argv: List[str]) -> argparse.Namespace:
     parser.add_argument("-r", dest="ref", default="")
     parser.add_argument("--resume-from", dest="resume_from", default=None)
     parser.add_argument("--plan", dest="plan_source", default=None)
+    parser.add_argument("--spec", dest="spec_path", default=None)
     parser.add_argument("--model", dest="model", default=None)
     parser.add_argument("instructions", nargs="*")
     args = parser.parse_args(argv)
```
[the remark this answers is not in the document]
Comment (xbrianh) [reply gremlins/orchestrators/gh.py:94]: Fixed — updated the usage string in  to include .

File: gremlins/orchestrators/local.py [hunk 4]
```diff
@@ -93,6 +93,7 @@ def _parse_local_args(argv: List[str]) -> argparse.Namespace:
     parser.add_argument("--resume-from", dest="resume_from", default=None,
                         choices=VALID_RESUME_STAGES)
     parser.add_argument("--plan", dest="plan_path", default=None)
+    parser.add_argument("--spec", dest="spec_path", default=None)
     parser.add_argument("instructions", nargs="*")
```
[the remark this answers is not in the document]
Comment (xbrianh) [reply gremlins/orchestrators/local.py:97]: Fixed — updated the usage string in  to include .

File: gremlins/orchestrators/gh.py [hunk 5]
```diff
@@ -389,6 +402,7 @@ def stage_implement() -> None:
             is_git=True,
             kind="gh",
             issue_num=issue_num,
+            spec_text=spec_text,
         )
```
[the remark this answers is not in the document]
Comment (xbrianh) [reply gremlins/orchestrators/gh.py]: Fixed — wrapped the read in try/except (OSError, UnicodeDecodeError); emits a warning to stderr and proceeds with spec_text="" on failure.

File: gremlins/orchestrators/local.py [hunk 6]
```diff
@@ -228,6 +241,7 @@ def stage_implement() -> None:
             core_principles=core_principles,
             session_dir=session_dir,
             is_git=is_git,
+            spec_text=spec_text,
```
[the remark this answers is not in the document]
Comment (xbrianh) [reply gremlins/orchestrators/local.py]: Fixed — wrapped the read in try/except (OSError, UnicodeDecodeError); emits a warning to stderr and proceeds with spec_text="" on failure.

File: gremlins/stages/implement.py (hunk 1 above)
Comment (xbrianh) [reply gremlins/stages/implement.py:83]: Done — truncation now finds the last newline before the 50000-char limit with rfind and falls back to the hard cut only when there is no newline in range. Added a test for the newline-boundary case.

File: gremlins/orchestrators/gh.py (hunk 2 above)
Comment (xbrianh) [reply gremlins/orchestrators/gh.py:303]: Added a comment explaining the guards exist for direct (non-launcher) invocations since launcher.py normalizes spec_path before spawning. Applied the same comment to local.py for consistency.
